cpumf/man/chcpumf.8
.\" chcpumf.8
.\"
.\"
.\" Copyright IBM Corp. 2014, 2017
.\" s390-tools is free software; you can redistribute it and/or modify
.\" it under the terms of the MIT license. See LICENSE for details.
.\" ----------------------------------------------------------------------
.TH chcpumf "8" "February 2014" "s390-tools" "CPU-MF management programs"
.
.ds c \fBchcpumf\fP
.
.
.SH NAME
chcpumf \- manage the CPU-measurement facilities support
.
.
.SH SYNOPSIS
.B chcpumf
.RB [ \-m | \-\-min
.IR NUM_SDB ]
.RB [ \-x | \-\-max
.IR NUM_SDB ]
.RB [ \-V |  \-\-verbose ]
.br
.B chcpumf
.BR \-h | \-\-help
.br
.B chcpumf
.BR \-v | \-\-version
.
.
.SH DESCRIPTION
The chcpumf program helps you to manage settings for the CPU-measurement
facilities for Linux on System z.
.
.
.SH OPTIONS
.TP
.BR \-m ", " \-\-min " \fINUM_SDB\fP"
Specifies the minimum sampling facility buffer size in sample-data-blocks
(SDB).  A sample-data-block consumes about 4 kilobytes.

This is the initial buffer size when you start the sampling facility.
.
.TP
.BR \-x ", " \-\-max " \fINUM_SDB\fP"
Specifies the maximum sampling facility buffer size in sample-data-blocks
(SDB).  A sample-data-block consumes about 4 kilobytes.

This is the maximum size to which the buffer is dynamically adjusted during
sampling.
.
.TP
.BR \-V ", " \-\-verbose
Displays verbose messages.
.
.TP
.BR \-h ", " \-\-help
Displays help information, then exits.
.
.TP
.BR \-v ", " \-\-version
Displays version information, then exits.
.
.
.SH "SEE ALSO"
.BR lscpumf (1)


CHANGELOG.md
Release history for s390-tools (MIT version)
--------------------------------------------

* __v2.39.0 (2025-10-14)__

  For Linux kernel version: 6.16 / 6.17

  Changes of existing tools:
  - chpstat: Add options to select IEC units for scaling (SI units are default)
  - chzdev: Introduce --no-module-load option
  - cpi: Disable CPI for SEL guests by default
  - dbginfo.sh: Enhance logging on timeout triggered
  - iucvterm: Install symlink for lsiucvallow.8 man page
  - lshwc: Add command line flag to specify individual counters
  - lspai: Add command line flag for delta values
  - lspai: Add command line flag for short counter names
  - lspai: Add command line flag to specify individual counters
  - lspai: Add command line flags for all cpus
  - lspai: Add command line flags for hexadecimal output
  - man: Use CR for constant width font
  - pvimg: Add '--image-key' option
  - zdev: Allow dynamic control of module load
  - zipl/boot: Fix EBCDIC code page 500 conversion and decrease size by 200 bytes
  - zipl: Add support of heterogeneous mirrors (remove technical limitations
    on mirrored targets, thus allowing mirrored devices consist of partitions
    at different offsets on disks of different types and geometry).
  - zkey: Add support for generating and importing exportable secure keys

  Bug Fixes:
  - chpstat: Fix scaling of DPU utilization calculation
  - zdev/dracut: Prevent loading of unused kernel modules
  - zdev: Fix double device configuration on DPM systems
  - zdev: Fix double device configuration with rd.dasd
  - zipl_helper.device-mapper: Fix segfault in an error path

* __v2.38.0 (2025-06-25)__

  For Linux kernel version: 6.14 / 6.15

  Add new tools:
  - udev: New rule to set newly hotplugged CPUs online
  - zmemtopo: Display memory topology information
  - zpwr: Display power readings of a partition and CPC

  Removed tools / features:
  - check_hostkeydoc: Remove installation target
  - scsi_logging_level: Delete SCSI logging script (available in sg3_utils)
  - zdump: Drop build_arch for s390 DASD dumps
  - zdump: Drop non-extended multi-volume DASD dump support
  - zdump: Drop support of 32-bit dump architecture
  - zdump: Drop support of non-extended single volume DASD dumpers
  - zdump: Drop support of obsolete dumps and dumpers

  Changes of existing tools / libraries:
  - Various man-pages fixes
  - check_hostkeydoc: Add deprecation warning
  - check_hostkeydoc: Move to scripts directory
  - cpuplugd: Allow cpu hotplugging on systems without polarization
  - dbginfo.sh: Add Ubuntu snap tool
  - dbginfo.sh: Add missing config data and logs
  - dbginfo.sh: Reworking the container section
  - dbginfo.sh: Update for network commands
  - dbginfo.sh: Updating info for disks and lvm
  - libutil: Add machine type definition for machines 9175 and 9176
  - lscpumf: Add support for IBM z17 counter sets
  - lshwc: Add command line flag for run time
  - lshwc: Add flags to display counter values in hex
  - lshwc: Add output '--format' option
  - lshwc: Add support for delta counter value display
  - lspai: Add output '--format' option
  - lsreipl: Add secure boot state to output
  - lswhc: Add short names to lshwc output
  - pv_tools: Add Bash and Zsh completions
  - pvapconfig: Add '--unbind' option
  - pvimg/boot: Print error messages from stage3a bootloader
  - pvimg: Add support for CCK update
  - pvsecret: Add support for CCK update
  - pvsecret: Allow retrieving secrets by index & warn for duplicated entries
  - pvsecret: Deny adding secrets with duplicated secret IDs
  - zdev: Add support for virtio devices
  - zipl: Enhance mirror support
  - zipl: Implement '--dry-run' option for all dump jobs
  - zipl_helper.device-mapper: Support mirrors over NVMe devices
  - zkey/dracut: Add a dracut config file for zkey
  - zkey/initramfs: Update initramfs hook to correct drivers and include zkey plugins
  - zkey: Add support for converting a clear-key LUKS2 volume to use a secure key

  Bug Fixes:
  - chpstat: Add missing CMG 5 data fields
  - chpstat: Fix DPU utilization calculation
  - libutil/util_file: Handle over-read in util_file_read_fd()
  - pvattest: Fix successful 'check' evaluation
  - pvsecret: Fix some edge cases for plaintext keys
  - zipl_helper.device-mapper: Fix imprecise is_device_mapper() predicate
  - zkey: Fix EP11 secure key reencipher function
  - zpcictl: Fix command line parsing for invalid options

* __v2.37.0 (2025-02-07)__

  For Linux kernel version: 6.13

  Changes of existing tools:
  - dbginfo.sh: Add details on CPU-measurement
  - dbginfo.sh: Add new crypto command
  - dbginfo.sh: Add overview commands and crypto update
  - dbginfo.sh: Adding kdump info
  - dbginfo.sh: Removing outdated email references
  - dbginfo.sh: Rework network section
  - dbginfo.sh: Update copyright 2nd year
  - pvimg: Add '--(enable|disable)-image-encryption' flags to 'pvimg create'
  - pvimg: Add '--cck <FILE>' command line option and make '--comm-key' an alias
  - pvimg: Add '--hdr-key' command line option to 'pvimg create'
  - pvimg: Rename '--key' into '--hdr-key' and use '--key' as an alias (for 'pvimg info')
  - pvsecret: Add support for retrievable secrets
  - ziorep_config: Add PCHID field to adapter report
  - ziorep_traffic: Add DEVBUSID column to traffic report
  - ziorep_utilization: Add --fcp-device parameter to print virtual adapter report
  - ziorep_utilization: Add PCHID column to physical adapter report
  - ziorep_utilization: Now prints only physical adapter report by default
  - ziorep_utilization: Swap Bus-ID and CHPID columns in virtual adapter report
  - zipl/boot: Increase section size for eckd_mv dumper
  - zkey: Add support for listing and importing protected virtualization secrets

  Bug Fixes:
  - chpstat: Fix invalid utilization data on older kernels
  - opticsmon: Fix runaway loop in on_link_change()
  - zipl: Update inline assembly for GCC 15
  - zipl_helper.device-mapper: Add missed step in logical device resolution


cpumf/chcpumf.c
/*
 * chcpumf -  Change CPU Measurement Facility Characteristics
 *
 * Copyright IBM Corp. 2020, 2022
 *
 * s390-tools is free software; you can redistribute it and/or modify
 * it under the terms of the MIT license. See LICENSE for details.
 */

#include <err.h>
#include <fcntl.h>
#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/stat.h>

#include "lib/libcpumf.h"
#include "lib/util_opt.h"
#include "lib/util_path.h"
#include "lib/util_prg.h"

static unsigned int verbose;
static unsigned long min_sdb, max_sdb;

static struct util_opt opt_vec[] = {
	UTIL_OPT_SECTION("OPTIONS"),
	{
		.option = { "min", required_argument, NULL, 'm' },
		.argument = "num_sdb",
		.desc = "Specifies the initial size of the sampling buffer.\n"
			"A sample-data-block (SDB) consumes about 4 kilobytes.",
	},
	{
		.option = { "max", required_argument, NULL, 'x' },
		.argument = "num_sdb",
		.desc = "Specifies the maximum size of the sampling buffer.\n"
			"A sample-data-block (SDB) consumes about 4 kilobytes.",
	},
	{
		.option = { "verbose", no_argument, NULL, 'V' },
		.desc = "Verbose, display new sample-data-block values.",
	},
	UTIL_OPT_HELP,
	UTIL_OPT_VERSION,
	UTIL_OPT_END
};

static const struct util_prg prg = {
	.desc = "Change CPU Measurement facility charactertics",
	.copyright_vec = {
		{
			.owner = "IBM Corp.",
			.pub_first = 2020,
			.pub_last = 2020,
		},
		UTIL_PRG_COPYRIGHT_END
	}
};

static long parse_buffersize(char *string)
{
	char *suffix;
	long bytes;

	bytes = strtol(string, &suffix, 10);
	if (strlen(suffix) > 1)
		return -1;
	switch (*suffix) {
	case 'k':
	case 'K':
		bytes *= 1024;
		break;
	case 'm':
	case 'M':
		bytes *= 1048576;
		break;
	case '\0':
		break;
	default:
		return 0;
	}
	return bytes;
}

static int write_sfb(unsigned int min, unsigned int max)
{
	int rc = EXIT_SUCCESS;
	char text[64], *path;
	size_t len;
	FILE *fp;

	path = util_path_sysfs(S390_CPUMSF_BUFFERSZ);
	fp = fopen(path, "w");
	if (!fp)
		err(EXIT_FAILURE, "%s", path);
	snprintf(text, sizeof(text), "%u,%u", min, max);
	len = strlen(text) + 1;
	if (fwrite(text, 1, len, fp) != len) {
		warn("%s", path);
		rc = EXIT_FAILURE;
	}
	if (fclose(fp)) {
		warn("%s", path);
		rc = EXIT_FAILURE;
	}
	if (verbose && rc != EXIT_FAILURE)
		warnx("Sampling buffer sizes:\n"
		      "    Minimum:%7d sample-data-blocks\n"
		      "    Maximum:%7d sample-data-blocks\n",
		      min, max);
	free(path);
	return rc;
}

static int parse_args(int argc, char **argv)
{
	int opt, action = 0;
	long new;

	while ((opt = util_opt_getopt_long(argc, argv)) != -1) {
		switch (opt) {
		case 'h':
			util_prg_print_help();
			util_opt_print_help();
			exit(EXIT_SUCCESS);
		case 'v':
			util_prg_print_version();
			exit(EXIT_SUCCESS);
		case 'x':
			new = parse_buffersize(optarg);
			if (new < 1)
				errx(EXIT_FAILURE,
				     "The specified number(s) are not valid");
			max_sdb = new;
			action = 1;
			break;
		case 'm':
			new = parse_buffersize(optarg);
			if (new < 1)
				errx(EXIT_FAILURE,
				     "The specified number(s) are not valid");
			min_sdb = new;
			action = 1;
			break;
		case 'V':
			verbose = 1;
			break;
		default:
			util_opt_print_parse_error(opt, argv);
			exit(EXIT_FAILURE);
		}
	}
	if (!action)
		errx(EXIT_FAILURE, "You must specify a valid option");
	return action;
}

int main(int argc, char **argv)
{
	unsigned long my_min, my_max;

	util_prg_init(&prg);
	util_opt_init(opt_vec, NULL);

	parse_args(argc, argv);
	if (geteuid())
		errx(EXIT_FAILURE, "Must run as root");
	if (!libcpumf_have_sfb())
		errx(EXIT_FAILURE,
		     "No CPU-measurement sampling facility detected");
	libcpumf_sfb_info(&my_min, &my_max);
	if (!min_sdb)
		min_sdb = my_min;
	if (!max_sdb)
		max_sdb = my_max;
	return write_sfb(min_sdb, max_sdb);
}
